File: app/src/main/java/com/example/prashant/databinding/di/module/ViewModelModule.java
package com.example.prashant.databinding.di.module;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import com.example.prashant.databinding.di.ViewModelKey;
import com.example.prashant.databinding.observer.ItemFactory;
import com.example.prashant.databinding.observer.ItemViewModel;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@SuppressWarnings("unused")
@Module
public abstract class ViewModelModule {
    @Binds
    @IntoMap
    @ViewModelKey(ItemViewModel.class)
    public abstract ViewModel bindItemViewModel(ItemViewModel viewModel);

    @Binds
    public abstract ViewModelProvider.Factory bindViewModelFactory(ItemFactory factory);
}

File: app/src/main/java/com/example/prashant/databinding/di/module/ViewModelModule.kt
package com.example.prashant.databinding.di.module

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.example.prashant.databinding.di.ViewModelKey
import com.example.prashant.databinding.observer.ItemViewModel
import com.example.prashant.databinding.observer.ViewModelFactory
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
abstract class ViewModelModule {
    @Binds
    @IntoMap
    @ViewModelKey(ItemViewModel::class)
    abstract fun bindItemViewModel(viewModel: ItemViewModel): ViewModel

    @Binds
    abstract fun bindViewModelFactory(factory: ViewModelFactory): ViewModelProvider.Factory
}